common/router.cc

```diff
@@ -0,0 +1,98 @@
+#include "router.h"
+#include <iostream>
+#include <thread>
+
+Router::Router(const std::vector<std::string>& bootstrap_servers) {
+    // Iterate over bootstrap servers to find a reachable one
+    for (const auto& server : bootstrap_servers) {
+        try {
+            auto channel = grpc::CreateChannel(server, grpc::InsecureChannelCredentials());
+            stub_ = message_queue::message_queue::NewStub(channel);
+            std::cout << "Connected to bootstrap server: " << server << std::endl;
+            break;
+        } catch (const std::exception& e) {
+            std::cerr << "Failed to connect to bootstrap server: " << server << " - " << e.what() << std::endl;
+        }
+    }
+
+    if (!stub_) {
+        throw std::runtime_error("Failed to connect to any bootstrap server");
+    }
+}
+
+std::string Router::GetLeader(const std::string& topic, int partition) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (routing_table_.count(topic) && routing_table_[topic].count(partition)) {
+        return routing_table_[topic][partition];
+    }
+
+    // If partition leader is not found for a paritcular topic then fetch metadata for that topic
+    std::cerr << "Leader not found for topic: " << topic << ", partition: " << partition
+              << ". Refreshing metadata..." << std::endl;
+
+    FetchMetadata(topic);
+    if (routing_table_.count(topic) && routing_table_[topic].count(partition)) {
+        return routing_table_[topic][partition];
+    }
+
+    throw std::runtime_error("Failed to find leader after metadata refresh");
+}
+
+int Router::GetTotalPartitions(const std::string& topic) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (topic_partitions_.count(topic)) {
+        return topic_partitions_[topic];
+    }
+    throw std::runtime_error("Partition count not available for topic: " + topic);
+}
+
+// Fetch metadata for a given topic only if it is not already cached
+// If using cached metadat then need to ensure that cache is updated periodically
+// Another option is to always fetch metadata from the server insted of checking if cache exists.
+void Router::UpdateMetadata(const std::string& topic) {
```

Comment:
@Pratik-Sanghavi @yadavpa1 What do you think about this logic? We can use cached info in `routing_table_` to check if we should fetch metadata for the topic from server of not. But we need to find a way to refresh cache or update cache if the leader address present is not a valid one.

Reply:
Commented out the cache checking logic. For now every time producer will fetch the meta data from server before sending produce request to the server.

Reply:
LGTM